ROFO-20 간단한 유저 도메인 및 시큐리티 설정, 필터 작성

Unagi-zoso opened the pull request:
### 개요
유저 도메인과 관계되는 도메인을 위해 빠르게 개발될 필요가 있어 `id`, `nickname` 을 가지는 테이블 및 Entity 클래스 를 작성해야한다.
개발 환경에서 로그인 기능을 완성하기 전 시큐리티 인가를 위한 간단한 패스 필터를 작성해야한다.

### 변경사항
- 생성시간, 최근수정시간 관련 부모 Entity 클래스 작성
- 유저 테이블 및 Entity 클래스 작성
- RepositoryTest 어노테이션 작성
- 유저 Repository 및 테스트 작성
- Spring Security 종속성 및 Config 클래스 추가
- MockTestFilter 작성
- MockTestFilter 테스트 용 API 및 테스트 작성
- ERD CLOUD 유저 테이블 작성

### 테스트
- 테스트 코드 추가여부 O

### 관련 지라 및 위키 링크
 - [JIRA](https://weit-2nd.atlassian.net/browse/ROFO-20) 
 
<img width="395" alt="스크린샷 2024-06-04 오전 12 50 17" src="https://github.com/Weit-2nd/roady-foody-server/assets/97292432/b315c77e-1243-4278-bb8a-ff1d1ca8c441">


### 리뷰어에게 하고 싶은 말
`Entity` 유효성 검증을 `Request` 같은 DTO 에서 맡을 것이냐 `Entity` 에서 맡을 것이냐 고민을 했습니다.
둘 중 하나를 골라라하면 **Request 에 맡기는 것이 효과적**이다 생각합니다.(예외발생시점, 라이브러리)
하지만 `Entity` 에서 유효성 검사를 하지 않으면 개발자 사이에서 오류를 범할 수도 있지 않을까 생각이 들었습니다. 
그렇다고 `둘 다 유효성`을 하자니 해당 제약에 변경이라도 생기면 **고쳐야 할 곳이 두 군데**나 생기니 
유지보수적 관점에서 좋지 않을 것 같습니다. 
그래서 닉네임의 경우 정규표현식을 통해서 유효성을 검사하는데 이 정보를 한 객체에 두어 변경 전파를 줄이도록 하였습니다.
`src/main/kotlin/kr/weit/roadyfoody/support/regex/ValidationRegex.kt`

수정의사 열려있으니 어떤 피드백이라도 환영합니다!

Unagi-zoso committed:
ROFO-20 BaseTimeEntity 및 JpaAuditing 설정 추가

기획 중 Update 가 없는 기능이 있어 BaseTimeEntity 를 분리하였습니다.

Unagi-zoso committed:
ROFO-20 쓰이지 않는 파일 제거

Unagi-zoso committed:
ROFO-20 flyway baseline-version 설정

Unagi-zoso committed:
ROFO-20 Kotest DataJpaTest 호환 위한 config 클래스 추가

Unagi-zoso committed:
ROFO-20 RepositoryTest 전용 어노테이션 작성

TestContainers, DataJpaTest 등 repository 테스트에 필수로 붙을만한 어노테이션을 정리하였습니다.

Unagi-zoso committed:
ROFO-20 users 테이블 ddl 작성

Unagi-zoso committed:
ROFO-20 User 엔티티 작성

Unagi-zoso committed:
ROFO-20 UserRepository 및 테스트 작성

`JPA` 기본 제공 메소드의 `Optional` 반환 타입이나 `nullable` 한 값을 `repository` 에서 확장함수를 통해 다룹니다.

Unagi-zoso committed:
ROFO-20 테스트 코멘트 수정

기존 코멘트는 DCI 적이지 않다 생각해 수정하였습니다.
메소드 별로 새로 구분하였습니다.

Unagi-zoso committed:
ROFO-20 SpringSecurity 종속성 추가

Unagi-zoso committed:
ROFO-20 시큐리티 관련 클래스 작성

Unagi-zoso committed:
ROFO-20 MockPassFilter 테스트 및 TestController 문서 추가

Unagi-zoso committed:
ROFO-20 User nickname 유효성 검사 로직 및 테스트 작성

Unagi-zoso committed:
ROFO-20 Swagger 문서 경로 시큐리티 허용

이전 `sandbox` 환경에서 문서를 제대로 불러오지 못하는 경우가 있어 `"/v3/api-docs/**"` 경로를 추가하였습니다.

jifrozen0110 commented on src/main/kotlin/kr/weit/roadyfoody/domain/BaseTimeEntity.kt:
```diff
@@ -0,0 +1,26 @@
+    lateinit var createdDateTime: LocalDateTime
+        protected set
+}
+
```
BaseTimeEntity와 BaseModifiableEntity를 나눈 이유가 무엇인가요?

Unagi-zoso replied:
가게 정보 리포트 , 리뷰 같은 경우 1.0 버전에선 수정 기능이 없는 것으로 알고있습니다.
(리뷰의 경우 이후 버전에서 수정 기능이 추가될 수 있습니다.)
이런 테이블에도 최근수정시간을 두는 건 의미적으로 맞지 않고 공간낭비적이다 생각을 하여 미리 나눠뒀습니다!

konempty replied:
크 요런 사소한 디테일 너무 마음에드네요 ㅋㅋㅋ

jifrozen0110 commented on src/main/kotlin/kr/weit/roadyfoody/domain/user/User.kt:
```diff
@@ -0,0 +1,29 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "USERS_SEQ_GENERATOR")
+    @Column(name = "id", columnDefinition = "NUMERIC(19, 0)", updatable = false, nullable = false)
+    val id: Long = 0L,
```
기본값으로 0L을 설정한 이유가 궁금합니다!

Unagi-zoso replied:
결국 ID 를 결정하는건 시퀀스의 밸류 입니다만 
불변성, Null 허용하지 않는 타입, 생성자 사용 시 id 를 설정하지 않아도 된다는 점을 고려하였습니다.

jifrozen0110 commented on src/main/kotlin/kr/weit/roadyfoody/domain/user/User.kt:
```diff
@@ -0,0 +1,29 @@
+    @Column(name = "id", columnDefinition = "NUMERIC(19, 0)", updatable = false, nullable = false)
+    val id: Long = 0L,
+    @Column(name = "nickname", nullable = false, unique = true)
+    var nickname: String,
```
nickname length 설정도 해주면 좋을 것 같습니다!

Unagi-zoso replied:
으악 이걸 놓치다니 감사합니다!

jifrozen0110 replied:
length가 48인 이유가 궁금합니다!

jifrozen0110 commented on src/main/kotlin/kr/weit/roadyfoody/domain/user/UserRepository.kt:
```diff
@@ -0,0 +1,17 @@
+import org.springframework.data.jpa.repository.JpaRepository
+
+fun UserRepository.getByUserId(userId: Long): User {
+    return findById(userId).orElseThrow { IllegalArgumentException("$userId ID 의 사용자는 존재하지 않습니다.") }
```
오 이런식으로 repository 내부에서 Null 체크가 가능하군요!
저희 error 전략도 정하면 좋을 것 같습니다.

예를 들어 지금같이 사용자가 존재하지 않습니다에 IllegalArgumentException를 내보낼지

아니면 customError를 정의하여서

```
public class UserNotFoundException extends RuntimeException {

  public UserNotFoundException(String message) {
    super(message);
  }
}
```

ExceptionHandler.kt


```
@ExceptionHandler(value = UserNotFoundException.class)
public ResponseEntity<ErrorResponse> handleUserNotFoundException(UserNotFoundException ex) {
  traceManager.doErrorLog(ex);
  val userNotFoundErrorCode = ErrorCode.USER_NOT_FOUND; // 새로운 에러 코드 정의 ( ErrorCode 클래스에 추가 )
  return ResponseEntity.status(userNotFoundErrorCode.httpStatus)
      .body(ErrorResponse.of(userNotFoundErrorCode, ex.getMessage()));
}
```

ErrorCode.kt
```
enum class ErrorCode(val httpStatus: HttpStatus, val code: Int, val errorMessage: String) {
    USER_NOT_FOUND(HttpStatus.NOT_FOUND, -10008, "사용자가 존재하지 않습니다."),
```
이런식으로 custom하게 에러를 하나하나 관리할지 같이 정해보면 좋을 것 같아요!

Unagi-zoso replied:
그렇네요. 반영하겠습니다! 감사합니다!

jifrozen0110 commented on src/main/kotlin/kr/weit/roadyfoody/security/filter/MockPassFilter.kt:
```diff
@@ -0,0 +1,34 @@
+import org.springframework.web.filter.OncePerRequestFilter
+
+@Component
+class MockPassFilter : OncePerRequestFilter() {
```
해당 클래스는 테스트, 개발환경에서 사용하는 권한 mock 정보라고 생각하면 될까요?

Unagi-zoso replied:
넵

jifrozen0110 replied:
혹시 배포시에는 걷어내는 코드일까요? 그렇다면 주석으로 표시하는것이 좋을 것 같아서요!

Unagi-zoso replied:
이 부분은 고민 중인게 
개발 진도가 나가다보면 자연적으로 걷어낼 기능인지
아니면 테스트 코드 작성 편의를 위해 테스트 용 필터로 역할을 축소할 것인지 생각하고 있습니다.

jifrozen0110 reviewed (commented):
저희 파일구조도 같이 생각해보면 좋을 것 같습니다.
현재는 controller service끼리 폴더가 되어있지만 
CQRS 구조를 사용하기 때문에 도메인 하나당 2개이상의 서비스가 생길텐데

그렇게 되면 파일구조를

```
src/
└── main/
      └── java/
        └── com/
            └── yourcompany/
                └── user/
                    ├── controller/
                    │   └── UserController.java
                    ├── service/
                    │   └── UserQueryService.java
                    │   └── UserCommandService.java
                    ├── domain/
                    │   └── User.java
                    ├── repository/
                    │   └── UserRepository.java
                    └── exception/ custom하게 에러를 생성하게 되면
                        └── UserNotFoundException.java

```

이런식으로 파일구조를 가져가는건 어떨까요??

jifrozen0110 commented:
혹시 api 주소를 /api/v1/test/filter 이렇게 가져간 이유가 있을까요?? test를 위한 가짜 유저를 만드는 api 여서 그런가요?

jifrozen0110 commented:
> Entity 유효성 검증을 Request 같은 DTO 에서 맡을 것이냐 Entity 에서 맡을 것이냐 고민

저도 이런 고민 많이하였는데! 저는 Entity에 유횽성 검증 로직을 넣고 dto에는 valid 어노테이션을 통해 검증하는걸 선호합니다.

Entity에 유효성 검증 로직을 맡기는 이유는 request에서 값을 가져와서 비즈니스로직을 통과하게 되면 값이 바뀔수도있는데
이런 경우를 주의하기 위해 안전하게 최종적으로 데이터를 받게되는 Entity에 유효성 검사를 하는 편입니다!

Unagi-zoso commented:
nickname 변경 시 유효성 검사를 놓쳤네요..

Unagi-zoso commented:
> user/
>                     ├── controller/
>                     │   └── UserController.java
>                     ├── service/
>                     │   └── UserQueryService.java
>                     │   └── UserCommandService.java
>                     ├── domain/
>                     │   └── User.java
>                     ├── repository/
>                     │   └── UserRepository.java
>                     └── exception/ custom하게 에러를 생성하게 되면
>                         └── UserNotFoundException.java

@jifrozen0110 
전 처음 해보는 패키지 구조(제공해주신 구조)라 해보고 싶네요. 혹시 이렇게 구조를 잡을 시 얻을 수 있는 장점이 있나요?

제가 작업하기 이전에 만들어진 부분이 있어서 이 부분은 다른 분과도 상의해보죠.

jifrozen0110 commented:
> 전 처음 해보는 패키지 구조(제공해주신 구조)라 해보고 싶네요. 혹시 이렇게 구조를 잡을 시 얻을 수 있는 장점이 있나요?

개인적으로 개발하면서 느낀 장점은 서비스가 커져 도메인과 클래스가 많아지면 관리가 쉽다는 점입니다.
controller나 service가 여러 클래스가 생기게 되면 한개의 depth의 폴더에서 원하는 클래스를 찾기 힘든 경우가 많은데
도메인별로 폴더를 나눠서 두개 depth로 관리하면 빠르게 원하는 클래스를 찾을 수 있습니다

저희는 CQRS 아키텍처를 도입하기 때문에
Service는 최소 한개가 생길것이고
그렇게 되면 많은 Service 클래스가 생길것이니(예시 식으로)
예시 ) 
service
   l- UserQueryService
   l- UserCommandService
   l- StoreQueryService
....

도메인별로 디렉토리를 나눠서 관리하는것이 좋지않을까 라는 생각으로 말씀드렸습니다!

konempty commented on build.gradle.kts:
```diff
@@ -34,6 +34,7 @@ dependencies {
     implementation("org.springframework.boot:spring-boot-starter-actuator")
     implementation("org.springframework.boot:spring-boot-starter-validation")
     implementation("org.springdoc:springdoc-openapi-starter-webmvc-ui:2.5.0")
+    implementation("org.springframework.boot:spring-boot-starter-security:3.3.0")
```
요기에 버전 안적어도 
> id("org.springframework.boot") version "3.3.0" 

요거에 의해 3.3.0버전으로 설치되지 않을까요??

Unagi-zoso replied:
버전을 명시하지 않으면 
혹시라도 서로 호환되지 않는 버전의 종속성들이 생기게 된다던가 그런 이야기를 들은 적 있어서 개인적인 관습으로 명시하고 있습니다.
이런 문제가 발생하지 않나요?

konempty replied:
@Unagi-zoso 
네넵 주열님이 생각하는 그 이슈가 버전명시하지않으면 발생하긴 하지만
지금은 경우 id("org.springframework.boot") version "3.3.0" 요걸 적는것 만으로 
org.springframework.boot 하위 라이브러리들은 전부 버전을 명시한 효과가 발휘되구요

저걸 적었을경우 
나중에 누군가 스프링 버전을 올리기 위해
id("org.springframework.boot") version "3.3.0"
요것만 수정하고
implementation("org.springframework.boot:spring-boot-starter-security:3.3.0")
요걸 수정하지 않게되면 오히려 버전 불일치로 인한 이슈가 발생할것 같아요 ㅋㅋㅋ

Unagi-zoso replied:
아. 그렇군요. 정말 중요한 피드백 감사합니다.

konempty commented on src/main/kotlin/kr/weit/roadyfoody/config/SecurityConfig.kt:
```diff
@@ -0,0 +1,45 @@
+        "/api/v1/test/success",
+        "/api/v1/test/error",
+        "/swagger-ui/**",
+        "/v3/api-docs/**",
```
앗 ㅋㅋㅋ 프로메데우스 actuator URL도 추가해주셔야해요

konempty replied:
참고로 URL은  "/actuator/prometheus" 입니다~!

Unagi-zoso replied:
아이고. 바로 추가하겠습니다.

konempty commented on src/main/kotlin/kr/weit/roadyfoody/domain/user/User.kt:
```diff
@@ -0,0 +1,29 @@
+class User(
+    @Id
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "USERS_SEQ_GENERATOR")
+    @Column(name = "id", columnDefinition = "NUMERIC(19, 0)", updatable = false, nullable = false)
```
columnDefinition은 안적어줘도 Long형이면 알아서 Oracle DB에선 NUMERIC으로 잡히는걸로 알고있는데
명시해준이유가 있을까요??

Unagi-zoso replied:
아 제가 이전에 혼자하던 프로젝트에서 코드를 가져왔는데
그 때 `hibernate ddl` 생성 기능과 `h2 DB` 를 같이 사용하고 있어서 그 호환을 위해 명시해뒀습니다. 걷어내는걸 잊었네요. 현재는 `flyway` 를 통해 테이블이 관리되니 이런 문제는 일어나지 않겠어요.
감사합니다.

konempty replied:
@Unagi-zoso 
이거 혹시 오댜를 따라하신거라면
오댜때는 테스트 DB로 H2DB를 쓰다보니 오라클의 자료형과 일치하지 못해서 강제로 명시해준걸로 기억하고 있어요 ㅋㅋㅋ

konempty replied:
아 이미 이전에 답변을 하셨었군요 ㅋㅋㅋ

konempty commented on src/main/kotlin/kr/weit/roadyfoody/domain/user/UserRepository.kt:
```diff
@@ -0,0 +1,17 @@
+}
+
+fun UserRepository.getByNickname(nickname: String): User {
+    return findByNickname(nickname) ?: throw IllegalArgumentException("$nickname 닉네임의 사용자는 존재하지 않습니다.")
```
ㅋㅋㅋㅋㅋㅋㅋㅋ 이거 보니까 오댜를 많이 따라하신게 느껴지네요
요게 호불호가 갈릴수 있는 방식인데 이번에도 적용하는것도 나쁘진 않네요
저는 선호하는데 불호하시는분들은 자주 사용되는 메소드는 괜찮은데
딱한번 쓸 메소드까지 전부 이렇게 만들어 두는걸 싫어하시더라구요 ㅋㅋㅋ

아마 getByNickname 이 메소드도 딱 한번 쓰이지 않을까 싶네요

Unagi-zoso replied:
코프링 공부할 때 너무 잘 이용하고 있어요! 
사용빈도에 따른 판단도 앞으로 고려해보겠습니다!

konempty commented on src/main/kotlin/kr/weit/roadyfoody/security/handler/CustomAuthenticationEntryPoint.kt:
```diff
@@ -0,0 +1,28 @@
+
+@Component
+class CustomAuthenticationEntryPoint : AuthenticationEntryPoint {
+    @Throws(IOException::class)
```
오잉 요걸 붙혀두신 이유가 있을까요??

Unagi-zoso replied:
빈에 의존하지도 않는데 붙인건 
뭔가 딱 하나 뿐이고 스프링 시큐리티 설정에서 사용되다보니 습관적으로 붙였네요. 
아무래도 제거하는게 좋겠네요.

konempty replied:
요거 제거가 안되었군요 ㅋㅋㅋ

konempty commented on src/main/kotlin/kr/weit/roadyfoody/security/handler/CustomAuthenticationEntryPoint.kt:
```diff
@@ -0,0 +1,28 @@
+    ) {
+        response.contentType = "application/json"
+        response.status = HttpServletResponse.SC_UNAUTHORIZED
+        jacksonObjectMapper().writeValue(
```
jacksonObjectMapper를 매번 생성하는게 아니라 하나를 만들어두고 재활용하는 방식이 좋을것 같아요!

Unagi-zoso replied:
이건 제가 잘 모르고 사용하였네요. 
해당 메소드가 매번 ObjectMapper 를 새로 생성하는 불편함을 방지하려고 나온 줄 알았습니다.

konempty commented on src/main/kotlin/kr/weit/roadyfoody/support/regex/ValidationRegex.kt:
```diff
@@ -0,0 +1,5 @@
+
+const val NICKNAME_REGEX_STR = "^[A-Za-z0-9가-힣]{6,16}$"
+val NICKNAME_REGEX = Regex(NICKNAME_REGEX_STR)
+const val NICKNAME_REGEX_DESC = "닉네임은 6자 이상 16자 이하의 한글, 영문, 숫자 여야 합니다."
```
요 Min과 Max값도 상수로 빼두면 많은곳에서 재활용할수 있을것 같네요 ㅋㅋㅋ

konempty commented on src/main/resources/db/migration/V1__create_users.sql:
```diff
@@ -0,0 +1,11 @@
+CREATE TABLE users (
+                       id NUMERIC(19, 0) NOT NULL,
+                       nickname VARCHAR2(48) NOT NULL,
+                       created_datetime TIMESTAMP(6)  NOT NULL,
+                       updated_datetime TIMESTAMP(6) NOT NULL
```
6번라인에서 NOT NULL 앞에 스페이스바가 2개들어갔네요

konempty replied:
그리고 뭔가 포멧팅이 이상한것 같은데 요 파일에서 crtl + alt + L (맥의 경우 Ctrl + option + L) 한번만 눌러봐주세요

Unagi-zoso replied:
Intellij sql dialect 가 기본으로 MYSQL 로 설정되어있어 문법 오류 발생으로 포매팅 대상에서 제외됐었네요. 지금은 수정하였습니다.

konempty commented on src/test/kotlin/kr/weit/roadyfoody/controller/TestControllerTest.kt:
```diff
@@ -0,0 +1,42 @@
+@ActiveProfiles("test")
+@Import(TraceManager::class, SecurityConfig::class, CustomAuthenticationEntryPoint::class)
+@WebMvcTest(TestController::class)
+class TestControllerTest(
```
요 테스트는 TestController를 걷어낼때 같이 걷어내야한다고
TODO 주석 추가해주시면 좋을것 같아요!

konempty commented on src/test/kotlin/kr/weit/roadyfoody/domain/user/UserRepositoryTest.kt:
```diff
@@ -0,0 +1,90 @@
+            }
+        }
+
+        describe("existsById 메소드는") {
```
오잉 ㅋㅋㅋ 요 메소드는 기본제공 메소드 아닌가요??
요걸 테스트하신 이유가 궁금합니다!

Unagi-zoso replied:
습관적으로 해버렸네요. 요 메소드는 지우겠습니다!

konempty commented on src/test/kotlin/kr/weit/roadyfoody/domain/user/UserRepositoryTest.kt:
```diff
@@ -0,0 +1,90 @@
+        val nonexistentNickname = "JohnDoe"
+
+        beforeEach {
+            givenUser = userRepository.save(User(nickname = "existentNick"))
```
앞으로 여러분들이 개발하시면서 유저와 여러 엔티티들을 엄청나게 빈번하게 생성하는 코드를 작성해야할텐데
그 코드를 매번 작성하는게 생각보다 귀찮단말이죠 ㅋㅋㅋㅋ
그래서 보통 디폴트로 엔티티를 정의해두고 그걸 가져다 쓰는 TestFixture라는 개념을 쓰는걸 많이 사용하는데
요번 기회에 도입해주셔도 좋을것 같아요!

찾아보셨는데 만약 뭔지 잘 이해가 안된다면 편하게 말씀해주세요 제가 추가해둘게요 ㅋㅋㅋ

Unagi-zoso replied:
제가! 꼭 해보고싶습니다!

jifrozen0110 replied:
TestFixture를 이전 프로젝트에서 사용한적이 있어 [참고문서](https://github.com/Compono-Team/I-backend/pull/60/files#diff-4ff8da933ddf9ec9f3676ea20ac354e4aa3b99c57888a6c723069cd57e40f088)차 올려두겠습니다..!

konempty replied:
@jifrozen0110 
우와 ㅋㅋㅋ 이미 적용해보신 경력이 있으시군요

@Unagi-zoso 
지언님껄 보고 참고하되 저렇게 Request를 엔티티로 바꾸는것도 좋고 
그냥 바로 엔티티로 선언하는것도 좋은데 
어떤 방식이 나을지는 고민해보시고 적용부탁드릴게요~!

Unagi-zoso replied:
대박!!! 감사합니다!

konempty reviewed (changes requested):
ㅋㅋㅋㅋ 첫 PR부터 변경사항이 웅장하네요
너무 좋습니다 👍 👍 
조금만 수정부탁드립니다~!

konempty commented:
오호 지언님이 추천해주신 도메인별로 묶는것도 좋구요 그 이유도 타당하다고 생각해요!
그래서 저렇게 묶는거에 저는 찬성하는데 다만 관점의 차이에 따라 구조가 달라져요!

예를들어 UserController는
**"User"** 의 Controller로 볼건지
User의 **"Controller"** 로 볼건지에 따라 2가지 방법으로 묶을수 있어요
요 두개는 갠취라 뭘하든 상관없는데 실제 개발하실분들 의견이 중요하니까 개발자님들은 어떤걸 선호하시나요??

Case1. **"User"** 의 Controller (도메인 위주의 패키지 설계)
```
src/
└── main/
      └── java/
        └── com/
            └── yourcompany/
                └── user/
                    ├── controller/
                    │   └── UserController.java
                    ├── service/
                    │   └── UserQueryService.java
                    │   └── UserCommandService.java
                    ├── domain/
                    │   └── User.java
                    ├── repository/
                    │   └── UserRepository.java
                    └── exception/ custom하게 에러를 생성하게 되면
                        └── UserNotFoundException.java
                └── rivew/
                    ├── controller/
                    │   └── ....java
                    ├── service/
                    │   └── ....java
                    ├── domain/
                    │   └── ....java
                    ├── repository/
                    │   └── ....java
                    └── exception/ 
                        └── ....java

```

Case2. User의 **"Controller"** (레이어 위주의 패키지 설계)
```
src/
└── main/
      └── java/
        └── com/
            └── yourcompany/
                └── controller/
                    ├── user/
                    │   └──  UserController.java
                    └── review/
                        └──  ReviewController.java
                └── service/
                    ├── user/
                    │   └── UserQueryService.java
                    │   └── UserCommandService.java
                    └──  review/
                        └── ....java
                        └── ....java
                        └── ....java
                └── exception/ 
                    ├── user/
                    │   └── ....java
                    └── review/
                        └── ....java

```

PS. 요거 아스키 아트 하는것 같아서 예뻐보이는데 찍기가 너무 힘드네요 ㅋㅋㅋ큐ㅠ

Unagi-zoso committed:
ROFO-20 User 유효성 검사 포함한 닉네임 변경 메소드 작성

이렇게 만들어둬도 외부에서 그냥 `setter` 를 통해 변경할 수 있으니 이를 방지할 수 있는 방법이 없을까 고민하고 있습니다.

Unagi-zoso committed:
ROFO-20 커스텀 예외 관련 작업 및 적용

Unagi-zoso committed:
ROFO-20 User Entity 컬럼 속성 리뷰 반영

Unagi-zoso committed:
ROFO-20 ObjectMapper 싱글톤 작성 및 적용, 의미없는 스프링 빈 해제

Unagi-zoso committed:
ROFO-20 필터 미적용 경로 추가 및 필터, 핸들러 직접 생성

기존 필터, 핸들러는 스프링 빈에 의존하지 않아 더 이상 빈으로 관리하지 않는다. 따라서 직접 생성하도록 변경하였다.

Unagi-zoso committed:
ROFO-20 Spring Security 종속성 버전 제거

Gradle 상단 부에 Spring-boot 종속성에 대한 버전이 명시되어 있어 제거한다.

Unagi-zoso committed:
ROFO-20 닉네임 유효성 검사 요소 세분화 (글자 수 최소, 최대 길이 추가)

Unagi-zoso committed:
ROFO-20 유저 SQL 포매팅

기존엔 `Intellj` `SQL Dialect` 설정이 `Mysql` 로 되어있어 `Oracle` 문법을 인식하지 못하고 포매팅에서 제외시켜 보기가 힘들었다.
이 부분을 수정해 반영하였다.

Unagi-zoso committed:
ROFO-20 TestControllerTest 제거 타이밍 TODO 주석 작성

Unagi-zoso committed:
ROFO-20 UserFixture 추가 및 적용 , 부적절한 테스트 코드 제거

Repository 테스트를 할 땐 직접 정의한 메소드만을 테스트하였다.
이전 코드에는  `Hibernate` 에서 만들어주는 메소드까지 테스트하고 있어서 이를 제거하였다.

jifrozen0110 commented:
> Case1. **"User"** 의 Controller (도메인 위주의 패키지 설계)
> 
> ```
> src/
> └── main/
>       └── java/
>         └── com/
>             └── yourcompany/
>                 └── user/
>                     ├── controller/
>                     │   └── UserController.java
>                     ├── service/
>                     │   └── UserQueryService.java
>                     │   └── UserCommandService.java
>                     ├── domain/
>                     │   └── User.java
>                     ├── repository/
>                     │   └── UserRepository.java
>                     └── exception/ custom하게 에러를 생성하게 되면
>                         └── UserNotFoundException.java
>                 └── rivew/
>                     ├── controller/
>                     │   └── ....java
>                     ├── service/
>                     │   └── ....java
>                     ├── domain/
>                     │   └── ....java
>                     ├── repository/
>                     │   └── ....java
>                     └── exception/ 
>                         └── ....java
> ```
> 
> Case2. User의 **"Controller"** (레이어 위주의 패키지 설계)
> 
> ```
> src/
> └── main/
>       └── java/
>         └── com/
>             └── yourcompany/
>                 └── controller/
>                     ├── user/
>                     │   └──  UserController.java
>                     └── review/
>                         └──  ReviewController.java
>                 └── service/
>                     ├── user/
>                     │   └── UserQueryService.java
>                     │   └── UserCommandService.java
>                     └──  review/
>                         └── ....java
>                         └── ....java
>                         └── ....java
>                 └── exception/ 
>                     ├── user/
>                     │   └── ....java
>                     └── review/
>                         └── ....java
> ```
> 
> PS. 요거 아스키 아트 하는것 같아서 예뻐보이는데 찍기가 너무 힘드네요 ㅋㅋㅋ큐ㅠ

크게 상관없지만 저는 case.1 을 더 선호합니다!

konempty commented on src/main/kotlin/kr/weit/roadyfoody/support/jsonmapper/ObjectMapperProvider.kt:
```diff
@@ -0,0 +1,7 @@
+import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
+
+object ObjectMapperProvider {
+    val objectMapper = jacksonObjectMapper()
```
ㅋㅋㅋㅋ 요런식으로 싱글톤 운용하셔도 되는데
빈으로 선언해서 사용하는게 더 깔끔할것 같아요!

konempty commented on src/main/kotlin/kr/weit/roadyfoody/support/regex/ValidationRegex.kt:
```diff
@@ -0,0 +1,7 @@
+const val NICKNAME_MAX_LENGTH = 16
+const val NICKNAME_REGEX_STR = "^[A-Za-z0-9가-힣]{$NICKNAME_MIN_LENGTH,$NICKNAME_MAX_LENGTH}$"
+val NICKNAME_REGEX = Regex(NICKNAME_REGEX_STR)
+const val NICKNAME_REGEX_DESC = "닉네임은 6자 이상 16자 이하의 한글, 영문, 숫자 여야 합니다."
```
요기 
"닉네임은 ${NICKNAME_MIN_LENGTH}자 이상 ${NICKNAME_MAX_LENGTH}자 이하의 한글, 영문, 숫자 여야 합니다."
요렇게 상수를 재활용하면 어떨까요??

konempty commented on src/test/kotlin/kr/weit/roadyfoody/domain/user/UserTest.kt:
```diff
@@ -0,0 +1,81 @@
+        }
+
+        given("6 자 미만의 닉네임을 입력한 경우") {
+            val nickname = "12345"
```
으흠 ㅋㅋㅋ 여기에서도 NICKNAME_(MIN|MAX)_LENGTH에 의해 자동화 되도록 하면 좋지않을까요?? ㅋㅋㅋ
테스트 제목도 
"${NICKNAME_MAX_LENGTH}자 미만의 닉네임을 입력한 경우"
요렇게 바꾸보구요 ㅋㅋㅋ

좀 과한 수정인가 싶긴해서 좀 아니다 싶으면 요 코멘트는 넘어가셔도 좋습니다

konempty reviewed (approved):
저는 리뷰 다드린것 같구요
확인해보시고 적용가능한건 적용 부탁드리고
지언님이 말씀하신 패키지 구조만 세분이서 정해서 작업해주시면 저는 머지 해도 좋습니다~!
고생하셨습니다 👍 👍

jifrozen0110 reviewed (commented):
수고하셨습니다👍

Unagi-zoso commented:
@jifrozen0110 
일단 기획을 좀 의식했습니다.
```
서버개발자는 각 상황에 맞게 한글(3Byte)혹은 이모지(4Byte)를 기준으로 최대 글자수에 맞게 최대 글자수에 맞게 DB 컬럼을 잡아주세요
```

닉네임은 영어, 숫자, 한글로 가장 큰 바이트를 차지하는건 한글(3 Byte) 입니다.
그리고 닉네임은 최소 6, 최대 16자로 따라서 최대 크기를 3 * 16 -> 48 로 뒀습니다.

개인적으로도 이렇게 하는게 맞나 싶었는데 어떻게 생각하시나요?

jifrozen0110 commented:
아하! 검색해보니 오라클에서 한글은 3byte라고 하네요! 정말 꼼꼼하시네요. 잘봤습니다!

Unagi-zoso committed:
ROFO-20 ObjectMapperProvider 스프링 빈 등록 및 적용

Unagi-zoso committed:
ROFO-20 NICKNAME_REGEX_DESC 수정

Unagi-zoso committed:
ROFO-20 NICKNAME_LENGTH 참조 Test Fixture 작성 및 적용

Unagi-zoso commented:
패키지 설계 1번 안에 흥미가 가네요. 
도메인 별로 나눠서 개발하는 상황에서 얼마나 효과적일지 한 번 경험해보고 싶네요.

해당 PR 관련 질문, 피드백은 언제든지 받습니다.
